Migrate to Glide

pandasoft0 opened the pull request:
Replace Picasso library with Glide library. Issue: #1082

pandasoft0 commented:
Updated the code to use a Glide instead of Picasso.
Not sure about:
https://github.com/roma-apps/roma-android/blob/upstream/1082/migrate_to_glide/app/src/main/java/com/keylesspalace/tusky/util/CustomEmojiHelper.java
https://github.com/roma-apps/roma-android/blob/upstream/1082/migrate_to_glide/app/src/main/java/com/keylesspalace/tusky/fragment/ViewImageFragment.kt#L184-L212
https://github.com/roma-apps/roma-android/blob/upstream/1082/migrate_to_glide/app/src/main/java/com/keylesspalace/tusky/view/MediaPreviewImageView.kt#L101-L108

connyduck commented on app/src/main/java/com/keylesspalace/tusky/util/CustomEmojiHelper.java:
```diff
@@ -118,20 +123,19 @@ public void draw(@NonNull Canvas canvas, CharSequence text, int start, int end,
-        @Override
-        public void onBitmapFailed(Drawable errorDrawable) {}
+        Target<Bitmap> getTarget(){
+            return new SimpleTarget<Bitmap>() {
```
`SimpleTarget` is deprecated. I think we need `CustomTarget` here

connyduck commented on app/src/main/java/com/keylesspalace/tusky/adapter/StatusBaseViewHolder.java:
```diff
@@ -323,10 +325,10 @@ protected void setMediaPreviews(final List<Attachment> attachments, boolean sens
-                        .resize(maxW, maxH)
-                        .onlyScaleDown()
+                        .override(maxW,maxH)
+                        .downsample(DownsampleStrategy.CENTER_INSIDE)
```
This downsampling was used to prevent a crash when loading huge images. It looks like Glide is already resizing to the size of the ImageView, no need to help it

pandasoft0 replied:
I've removed downsampling

connyduck commented on app/src/main/java/com/keylesspalace/tusky/view/MediaPreviewImageView.kt:
```diff
@@ -94,18 +98,16 @@ defStyleAttr: Int = 0
-    // We do not handle the error here, instead it will be handled higher up the call chain.
-    override fun onError() {
+    override fun onResourceReady(resource: Drawable?, model: Any?, target: Target<Drawable>?, dataSource: DataSource?, isFirstResource: Boolean): Boolean {
+        onSizeChanged(width, height, width, height)
```
looks fine

pandasoft0 commented:
@connyduck 
1. I fixed the crash
2. But i see an ugly side effect and i have no ides how to solve it.
How to reproduce:
1. Upload a large image (3-4Mb)
2. Click on the image at the TimelineFragment while it doesn't loaded for preview yet.
a) at the Picasso version of the app it stucks until image load and starts the transition after
b) i updated this behaviour at the Glide version. It starts the transition immediately but show the progress bar until image load at the image preview screen. When you click "back" it start the exit transition, but shows the black placeholder at the timeline fragment. I think, Glide has not cached preview for this size and loads the image again, but i am not sure. Do you have any suggestions how to remove this side effect? I think, custom cache signature may help (https://bumptech.github.io/glide/doc/caching.html#cache-invalidation) What do you think?

connyduck commented:
The code in `ViewImageFragment` is such a mess. 
Here is how I refactored it and it seems to work and fixes #968:

- delete the whole `if (arguments!!.getBoolean(ViewMediaFragment.ARG_START_POSTPONED_TRANSITION)) {` block including the `else`.
- call `loadImageFromNetwork` instead 
- remove `.skipMemoryCache(true)` and  `.diskCacheStrategy(DiskCacheStrategy.NONE)` (**just let Glide decide whether to load the image from network or cache** - this seems to be the main problem here)
- remove `Handler().post {` (why is this even here)
- inline the `loadImageFromNetwork`
- do `photoActionsListener.onBringUp()` after the Glide call and remove it from `finishLoadingSuccessfully` (which could also be inlined)

- probably the whole `ARG_START_POSTPONED_TRANSITION` can also be removed from `ViewVideoFragment` but that requires some more investigation.

pandasoft0 commented:
> remove Handler().post { (why is this even here)
I have added this line to fix the crash. Glide crashes when app starts the new image load request to the same target from another callback.
I will test the app at my devices in few hours

connyduck commented:
Ahh yes makes sense. So it should be no longer needed.

charlag commented:
Please don't do Single.create(). fromCallaback should be used instead.
Original idea with disk loading was to know if transition should be instant or not. I cannot look at the code properly as I still lack proper connection.

connyduck commented:
Yes this was the idea, but it does not work as #968 shows. So the better option imho is to just transition always and show the ProgressBar if the image is not yet loaded.

pandasoft0 commented:
> Please don't do Single.create(). fromCallaback should be used instead.

Done

pandasoft0 commented:
> Here is how I refactored

Did you commit your changes or it's just your thoughts? If thoughts - i will update the code today

connyduck commented:
I tried it out but I discarded the changes afterwards. Yes, please update the code and resolve the merge conflicts. I will then test it again to see if its ready to merge.

pandasoft0 commented on app/src/main/java/com/keylesspalace/tusky/fragment/ViewImageFragment.kt:
```diff
@@ -74,53 +85,16 @@ class ViewImageFragment : ViewMediaFragment() {
-        }
+        loadImageFromNetwork(url, photoView)
 
+        //Add a delay to prevent screen blink on load image from the cache
```
@connyduck I was need to do this trick to prevent blinking. See video below

pandasoft0 commented:
Without a trick
https://www.dropbox.com/s/rim21jxczqtbt5h/without_trick.mp4?dl=0

With a trick
https://www.dropbox.com/s/llzpak0jq8ud2y1/with_trick.mp4?dl=0

pandasoft0 commented:
My thoughts.
When image cached at the memory - Glide fast enough to load it to the image view before transition, but if the image is cached at the local storage - app starts the transition before image load, it doesn't see the image and show activity without transition.

connyduck commented:
I think its not a good idea to wait with showing the Activity. This feels like lag. I rather have the activity open immediately than the animation.
Can we do something like this:
Check if the large image is in the memory cache, if yes, show it. If not, check if the preview image is in the memory cache, if yes, show it (maybe below the ProgressBar) and load the large image in the background. If the preview is not in memory cache, show the ProgressBar and load the large image in the background.

pandasoft0 commented:
I am not sure, we can check is image cached or not. I see a single way only - as it was done before - load image with cache=memory only and load from Network on fail. We can postpone transition and load image from any cache. If fail - show progress bar and load from internet...
Maybe, we can use custom cache implementation (I didn't check the ability) but it looks too complicated for this task.

connyduck commented:
My first preference is still just one single Glide call (basically everything you have now without the `Completable.timer`). Its simple and I think having no transition when loading from disk cache is ok for now.

> I am not sure, we can check is image cached or not. I see a single way only - as it was done before - load image with cache=memory only and load from Network on fail.

Yes you check if the image is cached by trying to load it from cache only. If you want to have an animation when loading from disk cache you have to load from any cache and not memory only. If you get everything to work this way its great, but we can also add it later.

pandasoft0 commented:
Please, check this new solution. It uses the 2 Glide requests but in the "Glide" documented way:
https://bumptech.github.io/glide/doc/options.html#starting-a-new-request-on-failure

https://github.com/roma-apps/roma-android/blob/ecc6169be793355833aecd4a7510f9608763dd1a/app/src/main/java/com/keylesspalace/tusky/fragment/ViewImageFragment.kt#L135-L149
